### ISSUE TITLE
prefixIds: { true } generates prefix__prefix__cls-1 class names

### ISSUE THREAD

**Issue by patrick-wc** [svg/svgo#1148]
Thanks very much to @strarsis for creating the plugin, I was having to manually change class names before using this which was awful. However when I use this plugin with gulp-imagemin I can't get the filename to be passed into the output svg filename. 

If I use this config (which from reading around this repo seems to be the way that is suggested, there is no documentation).

```
.pipe(
        imagemin([
          imagemin.gifsicle({ interlaced: true }),
          imagemin.jpegtran({ progressive: true }),
          imagemin.optipng({ optimizationLevel: 5 }),
          imagemin.svgo({
            plugins: [
              { removeViewBox: true },
              { cleanupIDs: false },
              { prefixIds: true }
            ]
          })
        ])
      )
```

The output file SVG class names (or IDs) are changed from "cls-1" to "prefix__prefix__cls-1" or in one file I noticed "Layer_2" being changed to "prefix__prefix__prefix__Layer_2"

If I use it with this config, (prefix part copied from [@ming-codes](https://github.com/svg/svgo/issues/674#issuecomment-328774019) )

```
.pipe(
        imagemin([
          imagemin.gifsicle({ interlaced: true }),
          imagemin.jpegtran({ progressive: true }),
          imagemin.optipng({ optimizationLevel: 5 }),
          imagemin.svgo({
            plugins: [
              { removeViewBox: true },
              { cleanupIDs: false },
              {
                prefixIds: {
                  prefix: {
                    toString() {
                      this.counter = this.counter || 0;

                      return `id-${this.counter++}`;
                    }
                  }
                }
              }
            ]
          })
        ])
      )
```

The output described above changes from "cls-1" to "id-125__id-111__cls-1" and "id-391__id-373__Layer_2" which although slightly off means I don't get any issues with the same SVGs sharing the same class names which makes them all screw up when inlining them on a page.

It still seems like something is being passed twice as there are repeated values (`prefix__prefix__` and `id-#__id-#`) which is similar to [this issue](https://github.com/svg/svgo/issues/1133), but I don't think I'm using multipass. I'm not sure why, it's probably my mistake on the configuration arguments but this has taken me a while to get working so I'm hoping this helps someone else. 

I am using the latest versions of gulp (4.0.2) and gulp-imagemin (6.1.0) on Windows 10 with Windows Subsystem Linux (WSL to help people out in future when searching).

Please let me know if I'm supposed to use prefixIds with some kind of function call other than true to get the filename coming through, I'm happy to provide more info if needed.

**Comment by strarsis** [comment]
@patrick-wc: The root of these issues is that there is no built-in plugin testing for CLI and plugin invocations. I have to set up a test case manually and try to reproduce this bug.

Just invoking the svgo CLI works fine:
`test.svg`:
```svg
<svg width="120" height="120" xmlns="http://www.w3.org/2000/svg">
    <rect class="test" x="10" y="10" width="100" height="100"/>
    <rect class="" id="test" x="10" y="10" width="100" height="100"/>
</svg>
````

````
./bin/svgo --enable=prefixIds -i ./test.svg -o test.min.svg
````
`test.min.svg`:
```svg
<svg width="120" height="120" xmlns="http://www.w3.org/2000/svg"><path class="test_svg__test" d="M10 10h100v100H10z"/><path d="M10 10h100v100H10z"/></svg>

````
The used prefix `test_svg__` looks fine, as the input file name is `test.svg` (dots (`.`) are replaced by underscores (`_`)).

**Comment by strarsis** [comment]
@patrick-wc: So I set up a test in my fork for `svgo` with `gulp-imagemin`:
````
$ git clone https://github.com/strarsis/svgo
$ cd svgo
$ git checkout gulp-test

($ nvm use lts/*) # if node version management is used

$ npm install
````

### Sorry, no better way for replacing `svgo` used by `imagemin-svgo` yet:
In file `node_modules/imagemin-svgo/index.js` replace 
`const SVGO = require('svgo');` on line 3 with 
`const SVGO = require('../..');`.
Save the changes.


````
$ cd gulp-test/
$ gulp
````

Inspect the svg files in `gulp-test/dist/images/` folder.

**Comment by strarsis** [comment]
@patrick-wc: So after some testing with that branch it seems that `gulp-imagemin` is the reason for the issue: `gulp-imagemin` passes the SVG content as buffer, without file path/name meta information:
https://github.com/sindresorhus/gulp-imagemin/blob/master/index.js#L78
`gulp-imagemin` used `imagemin` which in turn uses `imagemin-svgo` and it also just passes the SVG content as buffer, without file related meta data:
https://github.com/imagemin/imagemin-svgo/blob/master/index.js#L17
So `svgo` can't know the file path/name, hence it can't generate a prefix from it.

**Comment by wesselbaum** [comment]
Wouldn't it be sufficient checking if the prefix which will be added already exists in the unprefixed id/class?

**Comment by strarsis** [comment]
@wesselbaum: I still have to find out why there are SVGs are passed to `svgo` with prefixes. 
The SVGs should be only passed as original ones once.

**Comment by wesselbaum** [comment]
I won't disagree on this one for this issue. 
I've got similar use case, but I use svgo in my project for inline-SVGs and if there is a new SVG file I wanted to run one script which cleans up all SVGs. This approach won't work with the current prefixIds since after a couple runs the IDs will be very long. 
Long story short: Could you also consider my use case and maybe my solution?

**Comment by strarsis** [comment]
@wesselbaum: Yes, it seems reasonable. There is a catch though: What if, by chance, the classes start with something that should be also as the prefix? If in that case the prefix is skipped, there is no guarantee that there are no collisions. So what do you think about an extra option that can be still turned off, `skipPrefixed`, that would make the prefix plugin skip all classes/IDs/links that already start with the prefix to be applied?

**Comment by wesselbaum** [comment]
Sounds good to me. I appreciate your  work a lot. Thanks!

**Comment by strarsis** [comment]
@wesselbaum: So after I found out that `multipass` is indeed enabled for `imagemin-svgo`, 
a better fix emerged: Instead comparing the beginnings of the strings for the prefix to apply (which can result in collisions), an internal flag is simply added to each node processed by `prefixIds`.

You can try the fix out already without waiting it to be reviewed + merged 
by installing from its GitHub fork:
````
npm install github:strarsis/svgo#fix-prefixids-multipass
````
Depending on how your Gulp setup uses `svgo`, it may not override the `svgo` used by `gulp-imagemin`, `imagemin` and `imagemin-svgo`. You may have to go `node_modules/imagemin-svgo` and install the `svgo` there to override the `svgo` in `node_modules/svgo` inside.

**Comment by wesselbaum** [comment]
Sadly I can't try it until next week but I will report my results as soon as possible. Thanks.

**Comment by strarsis** [comment]
I just noticed that flagging each processed node isn't the best solution.
It would be better if `svgo` passes a multipass key to extra argument of each plugin.

**Comment by wesselbaum** [comment]
> You can try the fix out already without waiting it to be reviewed + merged
> by installing from its GitHub fork:
> 
> ```
> npm install github:strarsis/svgo#fix-prefixids-multipass
> ```
> 
The `skipPrefixed` solution is not available any more, right? The solution above doesn't work for my problem. You wrote that it could lead to collisions, but isn't it okay if it is an opt in solution?

**Comment by strarsis** [comment]
@wesselbaum: A new approach is used now that adds `multipassCount` to the `info` object that is passed to the plugins: https://github.com/svg/svgo/pull/1173
This allows the plugins to know about subsequent passes and skip them when necessary,
the `prefixIds` plugin uses that information in the PR to skip and prevent multiple prefixes.

````
npm install github:strarsis:add-multipass-info
````

**Comment by dphil** [comment]
@strarsis There appears to be a bug with #1173. The way multipassCount is initialized causes prefixIds to still run twice. I have proposed a fix in #1206.

**Comment by dphil** [comment]
Ah, I just found the open PR #1177 where you already caught this, @strarsis . I have closed mine as a duplicate.

**Comment by TrySound** [comment]
Fixed in 2.0.1
